vxturn/turn.py

```diff
@@ -0,0 +1,219 @@
+    else:
+        return ''
+
+
+def format_msisdn_for_whatsapp(msisdn):
+    return msisdn.lstrip("+")
+
+
+class TurnTransport(HttpRpcTransport):
+    CONFIG_CLASS = TurnTransportConfig
+
+    def respond(self, message_id, code, body=None):
+        if body is None:
+            body = {}
+
+        self.finish_request(message_id, json.dumps(body), code=code)
+
+    def send_message(self, message):
+        return yield treq.post(
+                self.config['outbound_url'],
+                self.get_send_params(message).to_json(),
+                headers={
+                    'User-Agent': ['Vumi Turn Transport'],
+                    'Content-Type': ['application/json'],
+                    'Authorization': ['Bearer {}'.format(self.config['token'])],
+                })
+
+    def get_send_params(self, message):
+        params = {
+            "preview_url": False,
+            "recipient_type": "individual",
+            "to": normalize_outbound_msisdn(message['to_addr']),
+            "type": "text",
+            "text": {"body": message['content']}
+        }
+
+        return params
+
+
+    @inlineCallbacks
+    def handle_raw_inbound_message(self, message_id, request):
+        try:
+            # TODO: validate HMAC secret
+
+            for message in request.args.get("messages", []):
+                content = ''
+                metadata = {}
+                if message['type'] == 'text':
+                    content = message["text"]["body"]
+                elif message['type'] == 'location':
+                    loc = message["location"]
+                    metadata["location"] = 'geo:{},{}'.format(loc['latitude'], loc['longitude'])
+
+                yield self.publish_message(
+                    transport_name='turn',
+                    transport_type='sms',
+                    message_id=message["id"],
+                    transport_metadata=metadata,
+                    to_addr=format_msisdn_for_whatsapp(self.config['to_addr']),
+                    from_addr=format_msisdn_for_whatsapp(message['from']),
+                    content=content,
+                    timestamp= iso8601(message["timestamp"]),
+                    )
+                log.msg("Inbound Enqueued.")
+
+            for event in request.args.get("statuses", []):
+                if event["status"] == 'sent':
+                    delivery_status = 'pending'
+                elif event["status"] == 'failed':
+                    delivery_status = 'failed'
+                elif status in ['delivered', 'read']:
+                    delivery_status = 'delivered'
+                else:
+                    continue
+
+                yield self.publish_delivery_report(
+                    user_message_id=event['id'],
+                    delivery_status=delivery_status,
+                    transport_metadata={
+                        'delivery_status': event["status"],
+                        'timestamp': iso8601(event['timestamp']),
```

Comment:
We can probably put this on the timestamp of the delivery report, instead of in the metadata